### PR TITLE
ExecAlias for foreign shells

### LINKED ISSUE #2898: Source-bash doesn't honor nested commands in aliases
A common syntax in bash is to pipe the output from one command into another, such as with the following bash alias:

```
alias gpo="git push --set-upstream origin \$(git rev-parse --abbrev-ref HEAD)"
```

Attempting to run that command after running source-bash on it results in this error:

```
$ gpo
error: unknown option `abbrev-ref'
usage: git push [<options>] [<repository> [<refspec>...]]
...
```

## xonfig

```
+------------------+-----------------+
| xonsh            | 0.7.9           |
| Python           | 3.7.1           |
| PLY              | 3.9             |
| have readline    | True            |
| prompt toolkit   | 2.0.4           |
| shell type       | prompt_toolkit2 |
| pygments         | None            |
| on posix         | True            |
| on linux         | False           |
| on darwin        | True            |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
+------------------+-----------------+
```

## Expected Behavior

Xonsh should honor this syntax or declare this syntax as unsupported.

## Replication Steps

```
# from any git repo
$ source-bash 'alias gpo="git push --set-upstream origin \$(git rev-parse --abbrev-ref HEAD)"'
$ gpo
```

### PR DESCRIPTION
**Pull request by scopatz**
This adds an exec alias for foreign shells, and performs refactors needed to make it clean. Closes #2898

### CODE REVIEW

**Comment by scopatz** [comment]
OK! This passes!

File: xonsh/foreign_shells.py [hunk 1]
```diff
@@ -433,36 +449,29 @@ def __init__(self, name, shell, filename, sourcer=None, extra_args=()):
...
+        if not hasattr(other, "_input_kwargs") or not callable(other._input_kwargs):
             return NotImplemented
-        return (
-            (self.name == other.name)
-            and (self.shell == other.shell)
-            and (self.filename == other.filename)
-            and (self.sourcer == other.sourcer)
-            and (self.extra_args == other.extra_args)
-        )
+        return self._input_kwargs == other._input_kwargs
```
Comment (melund) [inline xonsh/foreign_shells.py]: Shouldn't you call the functions here?

Comment (scopatz) [reply xonsh/foreign_shells.py]: Oh haha yeah :) Great catch @melund!

**Comment by melund** [comment]
Ok. I couldn't see others issues. Let others test it on master :)

**Comment by scopatz** [comment]
Thanks @melund!

**Pull request merged**